server/src/main/java/org/elasticsearch/common/xcontent/XContentHelper.java

```diff
@@ -575,4 +587,37 @@ public static void writeTo(StreamOutput out, XContentType xContentType) throws I
             out.writeVInt(xContentType.ordinal());
         }
     }
+
+    /**
+     * Construct {@link XContentFieldFilter} using given includes and excludes
+     * @param includes fields to keep, wildcard supported
+     * @param excludes fields to remove, wildcard supported
+     * @return filter using {@link XContentMapValues#filter(String[], String[])} if wildcard found in excludes, otherwise return filter using {@link XContentParser}
+     */
+    public static XContentFieldFilter newFieldFilter(String[] includes, String[] excludes) {
+        if ((CollectionUtils.isEmpty(excludes) == false) && Arrays.stream(excludes).filter(field -> field.contains("*")).count() > 0) {
+            return (originalSource, contentType) -> {
+                Function<Map<String, ?>, Map<String, Object>> mapFilter = XContentMapValues.filter(includes, excludes);
+                Tuple<XContentType, Map<String, Object>> mapTuple = XContentHelper.convertToMap(originalSource, true, contentType);
+                Map<String, Object> filteredSource = mapFilter.apply(mapTuple.v2());
+                BytesStreamOutput bStream = new BytesStreamOutput();
+                XContentType actualContentType = mapTuple.v1();
+                XContentBuilder builder = XContentFactory.contentBuilder(actualContentType, bStream).map(filteredSource);
+                builder.close();
+                return bStream.bytes();
+            };
+        } else {
+            final XContentParserConfiguration parserConfig = XContentParserConfiguration.EMPTY.withFiltering(
+                Set.of(includes),
+                Set.of(excludes)
+            );
+            return (originalSource, contentType) -> {
+                BytesStreamOutput streamOutput = new BytesStreamOutput(Math.min(1024, originalSource.length()));
+                XContentBuilder builder = new XContentBuilder(XContentType.JSON.xContent(), streamOutput);
```

Comment:
I think we should be using `contentType` rather than `XContentType.JSON` here? In practise we probably use json pretty much everywhere but there may be cases when the source is stored in CBOR or SMILE so we need to handle those cases as well.